File: Siftan.UnitTests/InListExpression_UnitTests.cs

namespace Siftan.UnitTests
{
  using System;
  using FluentAssertions;
  using NSubstitute;
  using NUnit.Framework;

  [TestFixture]
  public class InListExpression_UnitTests
  {
    [Test]
    [TestCase("A", true)]
    [TestCase("C", false)]
    public void IsMatch_DifferentScenarios_ReturnsExpectedResult(String term, Boolean expectedResult)
    {
      // Arrange
      InListExpression expression = new InListExpression(new[] { "A", "B" });
      Record record = new Record
      {
        Term = term
      };

      // Act
      Boolean result = expression.IsMatch(record);

      // Assert
      result.Should().Be(expectedResult);
    }

    [Test]
    public void HasReachedMatchQuota_SetToFirstMatchInList_ReturnsTrueAfterFirstMatch()
    {
      InListExpression expression = new InListExpression(new[] { "A", "B" }, InListExpression.MatchQuotas.FirstMatchInList);
      Record record = new Record
      {
        Term = "A"
      };

      // Act
      expression.IsMatch(record);

      // Assert
      expression.HasReachedMatchQuota.Should().BeTrue();
    }

    [Test]
    public void HasReachedMatchQuota_SetToFirstMatchOfEachInList_ReturnsTrueAfterAllTermsAreMatched()
    {
      InListExpression expression = new InListExpression(new[] { "A", "B", "C" }, InListExpression.MatchQuotas.FirstMatchOfEachTermInList);

      // Act
      expression.IsMatch(new Record { Term = "A" });
      Boolean firstResult = expression.HasReachedMatchQuota;
      expression.IsMatch(new Record { Term = "B" });
      Boolean secondResult = expression.HasReachedMatchQuota;
      expression.IsMatch(new Record { Term = "A" });
      Boolean thirdResult = expression.HasReachedMatchQuota;
      expression.IsMatch(new Record { Term = "C" });
      Boolean lastResult = expression.HasReachedMatchQuota;

      // Assert
      firstResult.Should().BeFalse();
      secondResult.Should().BeFalse();
      firstResult.Should().BeFalse();
      lastResult.Should().BeTrue();
    }

    [Test]
    public void HasReachedMatchQuota_SetToNone_ReturnsFalseAfterAllMatches()
    {
      InListExpression expression = new InListExpression(new[] { "A", "B", "C" });

      // Act
      expression.IsMatch(new Record { Term = "A" });
      Boolean firstResult = expression.HasReachedMatchQuota;
      expression.IsMatch(new Record { Term = "B" });
      Boolean secondResult = expression.HasReachedMatchQuota;
      expression.IsMatch(new Record { Term = "A" });
      Boolean thirdResult = expression.HasReachedMatchQuota;
      expression.IsMatch(new Record { Term = "C" });
      Boolean lastResult = expression.HasReachedMatchQuota;

      // Assert
      firstResult.Should().BeFalse();
      secondResult.Should().BeFalse();
      firstResult.Should().BeFalse();
      lastResult.Should().BeFalse();
    }
  }
}

File: Siftan.UnitTests/InListExpression_UnitTests.fs
namespace Siftan.UnitTests

open NUnit.Framework
open Siftan
open Jabberwocky.Toolkit.IO
open FsUnit

module ``InListExpression_UnitTests`` =

    [<Test>]
    [<TestCase("A", true)>]
    [<TestCase("C", false)>]
    let ``IsMatch returns expected result for different terms``(term: string, expectedResult: bool) =
        let exp = InListExpression([| "A"; "B" |])
        let record = Record(Term = term)
        exp.IsMatch(record) |> should equal expectedResult

    [<Test>]
    let ``HasReachedMatchQuota is true after first match in list``() =
        let exp = InListExpression([| "A"; "B" |], InListExpression.MatchQuotas.FirstMatchInList)
        let record = Record(Term = "A")
        exp.IsMatch(record) |> ignore
        exp.HasReachedMatchQuota |> should be True

    [<Test>]
    [<TestCase(InListExpression.MatchQuotas.FirstMatchOfEachTermInList, true)>]
    [<TestCase(InListExpression.MatchQuotas.None, false)>]
    let ``HasReachedMatchQuota is correct after all terms are matched at least once``(matchQuota: InListExpression.MatchQuotas, expectedResult: bool) =
        let exp = InListExpression([| "A"; "B"; "C" |], matchQuota)
        exp.HasReachedMatchQuota |> should be False

        exp.IsMatch(Record(Term = "A")) |> ignore
        exp.IsMatch(Record(Term = "B")) |> ignore
        exp.IsMatch(Record(Term = "A")) |> ignore
        exp.HasReachedMatchQuota |> should be False

        exp.IsMatch(Record(Term = "C")) |> ignore
        exp.HasReachedMatchQuota |> should equal expectedResult